### DISCUSSION TITLE
Setting Content-ID headers when composing multipart/related groups

### DISCUSSION THREAD

**Discussion by dcpurton** [neomutt/neomutt#3214] [Development]
It would be best for Neomutt to produce a RFC compliant `Content-ID` when sending email (accepting email is a different matter and Neomutt should be generous with what is accepted).

But there are some complexities:

- For the case of an automatically generated `Content-ID`, the domain name can be appended easily enough and this can be documented.
    - But this makes it slightly difficult for a script to predict the `Content-ID`, since it can change based on configuration.
    - Additionally the RFCs specify that the `Content-ID` should be *world* unique. This is pain if the user wants to reference the attachment, because not only can it not be predicted, but it will be long and ugly.
- For the case of a user specified `Content-ID`, what should happen if the user does not specify a RFC compliant `Content-ID`?
    - Should Neomutt change it?
        - But this is confusing and annoying for a user.
    - Should Neomutt warn?
        - Potentially adds configuration options on whether to warn or not and what should be done by default.
    - Should Neomutt just allow the user to do what they want?
- For the case of a resent/draft/postponed message containing a non-compliant `Content-ID` (or even a compliant one, but with the wrong domain), what should Neomutt do?
    - Change the `Content-ID` (as is done for the `Message-ID`)?
        - But this involves modifying attachments in the group, which seems like a bad idea.
    - Just use the `Content-ID` as is and ignore the fact that it isn't RFC compliant.

These complexities explain why I chose to break the specification in https://github.com/neomutt/neomutt/pull/3197 and just use the file name as the `Content-ID` when creating one automatically. It makes it easy to reference in the body of the email, easy to script, and the reality is that for Email, the `Content-ID` only *has* to be unique within the group.

Perhaps Neomutt should never automatically create or modify `Content-ID`s and just warn if they aren't present as is done for a lack of `Content-Language` when making a multilingual group. Then you can just document what a valid `Content-ID` is and hand over responsibility to the user.

**Comment by flatcap** [comment]
> produce a RFC compliant `Content-ID` when sending email
> but it will be long and ugly.

Yes.  I think something short and random,
Google's weren't too bad; the UUID ones to too long.

Your idea of using the filename seems good to start with -- the user can
anticipate the `Content-ID` -- but the moment a name has an awkward character
it breaks.

Whatever we choose, we need some way to communicate the ID to the user / script.

> the domain name can be appended

I don't like the sound of that

> Neomutt should be generous with what is accepted

Definitely

> `Content-ID` should be *world* unique

I don't quite understand their reasoning for this.

There were examples of creating a URL that specifies an Email using is `Message-ID`.
That clearly needs to be world-unique.  To reference a specific attachment, they
added the `Content-ID` to the end of the URL.

> For the case of a user specified `Content-ID`, what should happen if the user
> does not specify a RFC compliant `Content-ID`?

I think we should limit the `Content-ID` to `[a-zA-Z0-9_]+`
That gives the user plenty of scope for naming things sensibly.

If the user calls `<edit-content-id>`, we can create a sensible default.
The user can type over it if they wish.

> Should Neomutt change it?
> Should Neomutt warn?
> Should Neomutt just allow the user to do what they want?

I think we should fail if they don't meet our rules.
NeoMutt allows multi-line prompts, so we can put the rules on screen.

> For the case of a resent/draft/postponed message containing a non-compliant `Content-ID`
> Change the `Content-ID` (as is done for the `Message-ID`)?

If the `Content-ID` is meant to be world-unique, then it should be changed when
we resend an email, but it there are other attachments that refer to it, they'll
break.

So, perhaps we shouldn't touch existing `Content-ID`s.

**Comment by dcpurton** [comment]
Creating a default `Content-ID`, which the user can change subject to restriction seems sensible.

The code could be a simplified form of `gen_msgid` from `send/sendlib.c`:

```C
/**
 * gen_cid - Generate a unique Content ID
 * @retval ptr Content ID
 *
 * @note The caller should free the string
 */
static char *gen_cid(void)
{
  char rndid[MUTT_RANDTAG_LEN + 1];

  mutt_rand_base32(rndid, sizeof(rndid) - 1);
  rndid[MUTT_RANDTAG_LEN] = 0;

  return mutt_str_dup(rndid);
}
```
